Mage/src/main/java/mage/game/permanent/PermanentImpl.java

```diff
@@ -441,7 +441,15 @@ public Ability addAbility(Ability ability, UUID sourceId, Game game, boolean fro
         // other abilities -- any amount of instances
         if (!abilities.containsKey(ability.getId())) {
             Ability copyAbility = ability.copy();
-            copyAbility.newId(); // needed so that source can get an ability multiple times (e.g. Raging Ravine)
+
+            // needed so that source can get an ability multiple times (e.g. Raging Ravine)
+            // Deterministic to allow repeated ability copies to have same UUID
+            // (see https://github.com/magefree/mage/issues/10372)
+            long msb = ability.getId().getMostSignificantBits() + getId().getMostSignificantBits() + 1;
+            long lsb = ability.getId().getLeastSignificantBits() + getId().getMostSignificantBits() + 1;
```

Comment:
Can you explain how you came up with this algorithm?

Why "+1"?
Why add instead of bitwise XOR?
Why is lsb using msb from original id?

Reply:
> Can the test framework pick up the original issue? If so, can you include a test for it?

I will add a test that verifies that the propaganda copying is fixed, once i land on a solution that works and survives scrutiny.

> Also, does this change eliminate the need for the Whipgrass Entangler workaround?

In theory, it should (or more accurately, it will). I'll try removing that workaround and seeing if it still works as intended once i get the generalized solution complete.


> Can you explain how you came up with this algorithm?
> 
> Why "+1"? Why add instead of bitwise XOR? Why is lsb using msb from original id?

I might change it to an XOR. The +1 was a change i made on a hunch that didn't work out, and the MSB/LSB mixup is a genuine issue it seems. This is all going to need to be moved to CardUtil anyways, so this is pending some refactoring.